### PR TITLE
Node module support, window sizing, argument definitions

### PR DESCRIPTION
**Pull request by codecounselor**
This is a great project and I wanted to leverage it for some work I'm doing.  However, there were a couple of things I needed to add/fix.  Three of them represent open issues (#57, #73, #79).

I am not quite done but the changes have grown to the point I wanted to get this PR created so you can at least do some code review and know that these changes are out there.

### CODE REVIEW

File: charge.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+if (argv.version) {
+  console.log('v' + pkg.version)
+  process.exit(0)
+}
+
+if (argv.help || !input || !output) {
+  usage(1)
+} else {
+  exporter(input, output, argv)
```
Comment (codecounselor) [inline charge.js]: This is the fix for #57

File: index.js [hunk 2]
```diff
@@ -1,114 +1,138 @@
...
 
-var input = argv._[0] || argv.i || argv.input
-var output = argv._[1] || argv.o || argv.output
+var HTML_DPI = 96
 
-if (argv.v || argv.version) {
-  console.log('v' + pkg.version)
-  process.exit(0)
-}
+var PDFExporter = function (input, output, argv) {
```
Comment (codecounselor) [inline index.js]: Fixes #73 by getting the CLI/child process stuff out of the module entry point.  other file named `charge.js` because it "charges" a single electron export and then dies.  Someone could now use this module to create many exports inside the same process.

File: index.js [hunk 3]
```diff
@@ -1,114 +1,138 @@
...
+      'Letter': {x: 8.5 * HTML_DPI, y: 11 * HTML_DPI},
+      'Legal': {x: 8.5 * HTML_DPI, y: 14 * HTML_DPI}
     }
 
-    // if given a markdown, render it into HTML and return the path of the HTML
-    input = markdownToHTMLPath(input, opts, function (err, tmpHTMLPath) {
-      if (err) {
-        console.error('Parse markdown file error', err)
-        app.quit()
+    var pageDim = pageDimensions[pdfOptions.pageSize]
```
Comment (codecounselor) [inline index.js]: Fixes #79 by setting defaults based on pdf page size (I didn't add all dimensions yet) or by allowing height/width to be passed as options

File: charge.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+var parseArgs = require('minimist')
+var fs = require('fs')
+var path = require('path')
+var pkg = require('./package.json')
+var exporter = require('./index')
+
+var argOptions = {
```
Comment (codecounselor) [inline charge.js]: While looking into how to pass boolean arguments I saw that minimist supported these options and it looked a lot cleaner than all the conditions scattered about.

File: index.js [hunk 5]
```diff
@@ -1,114 +1,138 @@
...
+    pageDim = pdfOptions.landscape ? {x: pageDim.y, y: pageDim.x} : pageDim
+    var winX = argv.windowX || pageDim.x
+    var winY = argv.windowY || pageDim.y
+
+    console.info('Opening a browser window', winX, 'x', winY)
+    var win = new BrowserWindow({
+      width: winX,
+      height: winY,
+      x: 0, y: 0,
+      show: true,
```
Comment (codecounselor) [inline index.js]: Some of these options need cleaned up before you merge.  Accepting a hash of browser options is probably the way to go here because there are so many.  I saw a similar PR I believe.

File: index.js [hunk 6]
```diff
@@ -1,114 +1,138 @@
...
+      loadOpts.extraHeaders = 'pragma: no-cache\n'
+    }
+    win.loadURL(indexUrl, loadOpts)
+
+    win.webContents.on('did-finish-load', function () {
+      setTimeout(function () {
+        console.log('finished loading')
+
+        //  Image (PNG)
+        if (png) {
```
Comment (codecounselor) [inline index.js]: I need PNGS, figured others might too.  Activated if the output file ends with .png

**Comment by codecounselor** [comment]
@fraserxu - any feedback on this would be appreciated.  I'm trying to decide if going down the path of an official fork is the way to go here or not.  I think nearly all of my requirements are now included in this PR.

**Comment by fraserxu** [comment]
Thank you so much @codecounselor for the hard and great work! 

Sorry I was a bit busy recently but I'm more than happy to get this merged as soon as possible. I can't granntee that I'll merge it this week and I don't want make you to wait for me either to get your work done.

I'll try to take some time probably this week to review the code. Meanwhile I can add you as a collaborator after this if you are interested. It would be great to have a real user other than myself to contribute and mantain this project to make it better and healthy!

**Comment by codecounselor** [comment]
Thanks for the response.  I’d be happy to join as a collaborator.

I’m at the point now where I’m looking at all solutions for my use case which is to run this behind a node.js API.  Is that how you’re using it?  Do you have an opinion on the options available?

I have seen these projects, and a number of others as well
https://www.npmjs.com/package/electron-render-service
https://github.com/shankiesan/express-electron-pdf (which uses your project)
https://github.com/FWeinb/electron-screenshot-service

It seems like you must have been 1st to the party - I found your blog and electron-pdf at the top of my search results.  Thank you so much for that, its what led me down this path, which everyone on my team thinks is phenomenal.

I’m not sure how many people are actually using electron-pdf but it makes sense to me to invest in it since its so easy to find.  Do you know the owner of https://github.com/shankiesan/express-electron-pdf? 

I like the idea of keeping electron-pdf decoupled from the API Server, but the electron-render-service seems to have the best implementation of the bunch.  I have several questions out to the owner of that project which will likely determine the path I go from here.

It seems like using electron for exports is becoming very popular, but things are really sprawled out.  If you have contacts you’ve worked with I’d be interested in trying to pull things together into a single robust offering.

File: charge.js (hunk 4 above)
Comment (fraserxu) [inline charge.js]: Nice one!

File: package.json [hunk 7]
```diff
@@ -28,9 +28,11 @@
     "standard": "^6.0.6"
   },
   "dependencies": {
-    "electron-prebuilt": "^1.2.2",
+    "electron": "^1.4.1",
+    "electron-prebuilt": "^1.4.1",
```
Comment (fraserxu) [inline package.json]: I just came across this post and seems that `electron-prebuild` is deprecated and maybe we can simply keep the `electron` one?

**Comment by fraserxu** [comment]
@codecounselor I'm checking why [the test is failing](https://travis-ci.org/fraserxu/electron-pdf/jobs/164684300), and looks like test is failing because of 

> Starting with io.js 3 and Node.js 4, building native extensions requires C++11-compatible compiler, which seems unavailable on this VM. Please read https://docs.travis-ci.com/user/languages/javascript-with-nodejs#Node.js-v4-(or-io.js-v3)-compiler-requirements.

Add the following code to add the dependencies might able to fix the test, do you mind include this in the pull request so the test can pass? Thanks!

```yml
env:
  - CXX=g++-4.8
addons:
  apt:
    sources:
      - ubuntu-toolchain-r-test
    packages:
      - g++-4.8
```

File: index.js [hunk 8]
```diff
@@ -1,114 +1,193 @@
...
-  }
-}
+    var defaultOpts = {
+      width: pageDim.x,
+      height: pageDim.y,
+      enableLargerThanScreen: true,
+      show: false
+    }
+
+    var browserConfig = _.extend(defaultOpts, JSON.parse(argv.browserConfig || '{}'))
```
Comment (fraserxu) [inline index.js]: Might worth to add a `try catch` statement here to exist the process and throw an error message if user give an invalid JSON.

File: package.json (hunk 7 above)
Comment (codecounselor) [inline package.json]: Cool, that makes sense.  I will remove it

**Comment by codecounselor** [comment]
Comments addressed, merged with master, build is green.

File: package.json [hunk 9]
```diff
@@ -28,9 +28,10 @@
     "standard": "^6.0.6"
   },
   "dependencies": {
-    "electron-prebuilt": "^1.2.2",
+    "electron": "^1.4.3",
```
Comment (codecounselor) [inline package.json]: 1.4.3 was just released and among other things added new methods to `NativeImage`, one of which is `resize` which I am planning to use for scaling higher resolution images to fit on pptx slides.  Hint: creating powerpoints will be an upcoming PR, and I will likely add dependencies on `officegen` and a TBD library to convert pdf pages into images.  The reasoning is that PDF export does a better job with handling page breaks than can be done by image capture alone.

**Comment by fraserxu** [comment]
Great work @codecounselor 💯 🎉

File: package.json (hunk 9 above)
Comment (fraserxu) [inline package.json]: Let's see what we can do in another pr.

**Comment by fraserxu** [comment]
Published to v0.12.0 and added you as collaborator.
